promypy/main.py

```diff
@@ -100,29 +110,60 @@ def check(
     - If a file is not in the list, and is fully annotated, it will be ignored.
     - If a file is not in the list, and is not fully annotated, it will raise errors.
     """
+    args = shlex.split(mypy_args or "")
     with ignore_file.open("r") as file:
         all_ignored_files = {line.strip() for line in file.readlines()}
     files_to_ignore = set(files) & all_ignored_files
 
     files_with_error = set()
-    args = shlex.split(mypy_args or "")
-    result, _, exit_code = api.run(files + args)
-
-    output = []
-    if exit_code != 0:
-        for line in result.split("\n"):
-            match = re.match(FILE_PATTERN, line)
-            if ":" in line and match:
-                filename = match.group(1)
-                files_with_error.add(filename)
-                if filename not in files_to_ignore:
-                    output.append(line)
+    processedFiles = set()
+
+    with Progress(
+        *Progress.get_default_columns(),
+        TimeElapsedColumn(),
+        MofNCompleteColumn(),
+    ) as progress, ProcessPool() as pool:
+        total = len(files)
+        task = progress.add_task("Checking mypy on files...", total=total)
+        futures = [
+            pool.submit(call_mypy_api, args=[filename, *args], timeout=timeout)
+            for filename in files
+        ]
+
+        output = []
+        for future in as_completed(futures):
+            try:
+                filename, api_result = future.result()
+                result, _, exit_code = api_result
+                processedFiles.add(filename)
+                if exit_code == 2:
+                    output.append(f"Error in file: {filename}")
+                    files_with_error.add(filename)
+                    print(
+                        "promypy failed with exit status code 2. Please raise this issue with the developer."
+                    )
+                    print("Error details: ", api_result)
+                if exit_code != 0:
+                    for line in result.split("\n"):
+                        match = re.match(FILE_PATTERN, line)
+                        if ":" in line and match:
+                            filename = match.group(1)
+                            files_with_error.add(filename)
+                            if filename not in files_to_ignore:
+                                output.append(line)
+            except TimeoutError:  # pragma: no cover
+                progress.console.print("TimeoutError")
```

Comment:
Can we speed up the tool instead?

Reply:
That could be something done if they are going to continue using promypy for type checking purposes. This PR is just to have these changes on master in order to avoid using the forked repo.